### PR TITLE
iOS native scroll and feel

### PR DESCRIPTION
**Pull request by elijah-semyonov**
## Proposed Changes

Implement CupertinoScrollDecayAnimationSpec to simulate iOS inertia equation.
Implement CupertinoOverscrollEffect to simulate rubber banding and spring effect.

## Testing

Test: launch iOS app and test interaction between fling and overscroll

## Notes
Comments for CupertinoOverscrollEffect.direction mention this behavior for 2D cases:
Left picture shows visual path of combined scroll offset (scrollable content offset + overscroll offset), rect represents bounds of values `max(0, content.size - frame.size)` for each dimension.
Right pictures show value over time (t) for X and Y components. Green changes are happening inside `spring` animation, blue inside `fling` animation. (t) scale is matching for both (bottom and top on the right).
tempImagepBK72j
tempImageirkMZm

Currently the API forces us to communicate with outer entities using velocities deltas, which is built on assumption that all these animations' stages happen in sync for both dimensions (which is not the case).

```
override suspend fun applyToFling(
        velocity: Velocity,
        performFling: suspend (Velocity) -> Velocity
    ) {
        val availableFlingVelocity = playInitialSpringAnimationIfNeeded(velocity)

        val velocityConsumedByFling = performFling(availableFlingVelocity)
        val postFlingVelocity = availableFlingVelocity - velocityConsumedByFling

        playSpringAnimation(
            lastFlingUncosumedDelta.toFloat(),
            postFlingVelocity.toFloat(),
            false
        )
    }
```

### CODE REVIEW

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt [hunk 1]
```diff
@@ -90,6 +91,13 @@ internal val Animation<*, *>.durationMillis: Long
     get() = durationNanos / MillisToNanos
 
 internal const val MillisToNanos: Long = 1_000_000L
+const val SecondsToNanos: Long = 1_000_000_000L
```
Comment (MatkovIvan) [inline compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt]: internal? Or we can expose MillisToNanos too

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt [hunk 2]
```diff
@@ -90,6 +91,13 @@ internal val Animation<*, *>.durationMillis: Long
...
+
+fun convertSecondsToNanos(seconds: Float): Long =
+    (seconds.toDouble() * SecondsToNanos).roundToLong()
+
+fun convertNanosToSeconds(nanos: Long): Float =
```
Comment (MatkovIvan) [inline compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt]: Why not `Double`?

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * limitations under the License.
+ */
+
+package androidx.compose.animation.core
+
+import kotlin.math.abs
+import kotlin.math.ln
+import kotlin.math.pow
+
+/*
```
Comment (MatkovIvan) [inline compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt]: It's public class in common, it requires doc-comment

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/VectorizedAnimationSpec.kt [hunk 4]
```diff
@@ -722,6 +722,7 @@ class VectorizedTweenSpec<V : AnimationVector>(
     }
 }
 
+
```
Comment (MatkovIvan) [inline compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/VectorizedAnimationSpec.kt]: No changes in this file, please revert this line

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    DRAG, FLING
+}
+
+enum class CupertionOverscrollOffsetSpace {
+    LINEAR,
+    RUBBER_BANDED
+}
+
+@OptIn(ExperimentalFoundationApi::class)
+class CupertinoOverscrollEffect : OverscrollEffect {
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Should we move `Cupertino*` to separate package?

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/Overscroll.kt [hunk 6]
```diff
@@ -119,6 +122,17 @@ interface OverscrollEffect {
...
+
+    /**
+     * Represents an optional overscroll effect that can be applied when fling inertial movement crosses the overscroll area.
+     *
+     * @param scrollValueConverter an instance of [ScrollValueConverter] class which is used to convert the raw scroll offset
+     * into the overscroll effect offset.
+     * @return a nullable instance of the [FlingIntoOverscrollEffect] class, representing the overscroll effect to be used.
+     * If null, no effect will be applied.
+     */
+    fun flingIntoOverscrollEffect(scrollValueConverter: ScrollValueConverter): FlingIntoOverscrollEffect? =
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/Overscroll.kt]: New API in common won't be available on Android until upstreaming. 1.5 API already frozen, so we need to be careful about it and discuss such changes with Google

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt [hunk 7]
```diff
@@ -27,19 +30,16 @@ import androidx.compose.runtime.Stable
...
      * [androidx.compose.foundation.gestures.scrollable] that invoked this method.
-     *
+     * @param flingIntoOverscrollEffect The effect containing the logic executed
+     * if the fling inertial motion ends up in overscroll.
      * @return remaining velocity after fling operation has ended
      */
-    suspend fun ScrollScope.performFling(initialVelocity: Float): Float
-}
\ No newline at end of file
+    suspend fun ScrollScope.performFling(initialVelocity: Float, flingIntoOverscrollEffect: FlingIntoOverscrollEffect?): Float
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/FlingBehavior.kt]: Ouch. Such things will break bin compatibility with Android. Cannot be merged to 1.5

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt [hunk 8]
```diff
@@ -698,32 +689,75 @@ internal class DefaultFlingBehavior(
...
+                        val consumed = scrollBy(delta)
+                        lastValue = value
+                        velocityLeft = this.velocity
+
+                        val unconsumedDelta = delta - consumed
+
+                        // If some delta is not consumed, it means that fling hits into content bounds.
+                        // Unconsumed delta and current velocity will be initial values for
+                        // [flingIntoOverscrollEffect] animation to play if any, after we cancel decay animation
+                        if (abs(unconsumedDelta) > 0.5f) {
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt]: Please reuse `Float.isAboutZero()`

File: settings.gradle [hunk 9]
```diff
@@ -5,6 +5,7 @@ import org.gradle.nativeplatform.platform.internal.DefaultNativePlatform
 
 pluginManagement {
     repositories {
+        mavenLocal()
```
Comment (MatkovIvan) [inline settings.gradle]: Doesn't seem related

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 8 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt]: This particular line is reused from previous version of the code. I have a feeling that changing it could break things.

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt (hunk 2 above)
Comment (elijah-semyonov) [reply compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt]: Doesn't make sense indeed

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt (hunk 1 above)
Comment (elijah-semyonov) [reply compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/Animation.kt]: Can we? I'm trying not to change existing codebase.

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 8 above)
Comment (MatkovIvan) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt]: It's simple one. It wasn't used because I added `isAboutZero` recently and didn't change unrelated code

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt]: Oh, that's interesting. I thought `isAboutZero` to be about some really small epsilon. And was about to tell, that I didn't investigate accumulated error during delta consumption, so in case it doesn't fit that strict epsilon comparison, we could end up with having one extra frame of animation, where nothing visible happens (visual stutter from user experience side). But since it's exactly the same, I'd consider renaming `isAboutZero` to something like `isInsignificantForAnimation` and dual `isSignificantForAnimation` or at at least adding some comment to it. Which one you think is preferable?

File: settings.gradle (hunk 9 above)
Comment (elijah-semyonov) [reply settings.gradle]: Removed

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 8 above)
Comment (MatkovIvan) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt]: Let's avoid negation ("in") in names if possible

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 5 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: I think it's fine for now, especially considering they are in different packages (foundation/animation), same as the classes they are based on.

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import kotlin.math.pow
+
+/**
+ * A class that represents the animation specification for a scroll decay animation
+ * using iOS-style decay behavior.
+ *
+ * @property decelerationRate The rate at which the velocity decelerates over time.
+ * Default value is equal to one used by default UIScrollView behavior.
+ */
+class CupertinoScrollDecayAnimationSpec(
```
Comment (MatkovIvan) [inline compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt]: I guess we should move `Cupertino*` classes to separate package. Also please note that it will not work on android without upstreaming

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt [hunk 11]
```diff
@@ -571,49 +567,61 @@ private class ScrollingLogic(
...
+                result = result.update(
+                    performFling(available.toFloat().reverseIfNeeded()).reverseIfNeeded()
+                )
             }
         }
+
+        return result
+    }
+
+    suspend fun doFlingAnimationInNewScrollScope(available: Velocity): Velocity {
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt:618]: Should be private

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+    UNKNOWN, VERTICAL, HORIZONTAL
+}
+
+@OptIn(ExperimentalFoundationApi::class)
+class CupertinoOverscrollEffect(
+    /*
+     * Density to be taken into consideration during computations; Cupertino formulas use
+     * DPs, and scroll machinery uses raw values.
+     */
+    private val density: Float,
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Shouldn't we pass `Density` class here?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+
+    private var lastFlingUncosumedDelta: Offset = Offset.Zero
+    private val visibleOverscrollOffset: IntOffset
+        get() =
+            overscrollOffset.reverseHorizontalIfNeeded().rubberBanded().round()
+
+    override val isInProgress: Boolean
+        get() =
+        // If visible overscroll offset has at least one pixel
+            // this effect is considered to be in progress
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Formatting

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+//        val scrollValueConverter = scrollValueConverter ?: return velocity
+//
+//        val currentOverscroll = scrollValueConverter.convertOffsetToFloat(overscrollOffset)
+//
+//        return if ((velocity > 0f && currentOverscroll > 0f) || (velocity < 0f && currentOverscroll < 0f)) {
+//            playSpringAnimation(0f, velocity, true)
+//        } else {
+//            velocity
+//        }
+//    }
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Remove commented code

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+        when (this) {
+            NestedScrollSource.Drag -> CupertinoScrollSource.DRAG
+            NestedScrollSource.Fling -> CupertinoScrollSource.FLING
+            else -> null
+        }
+
+    /*
+     * Takes input scroll delta, current overscroll value, and scroll source. Returns pair of
+     * 1. Available delta to perform actual content scroll.
+     * 2. New overscroll value.
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Create 1-liner data class for such things

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+    ) {
+        val availableFlingVelocity = playInitialSpringAnimationIfNeeded(velocity)
+
+        val velocityConsumedByFling = performFling(availableFlingVelocity)
+        val postFlingVelocity = availableFlingVelocity - velocityConsumedByFling
+
+        playSpringAnimation(
+            lastFlingUncosumedDelta.toFloat(),
+            postFlingVelocity.toFloat(),
+            false
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]:
```suggestion
            flingFromOverscroll = false
```

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 17]
```diff
@@ -0,0 +1,405 @@
...
+        )
+    }
+
+    private fun Offset.toCupertinoOverscrollDirection(): CupertinoOverscrollDirection {
+        val epsilon = 1E-4f
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: const?
Is there anything to reuse in math utils?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+                CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+            }
+
+            CupertinoOverscrollDirection.HORIZONTAL -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+                CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+            }
+        }
+    private fun Velocity.toOffset(): Offset =
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: It doesn't look related to this class. Let's place such utils in file scope

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+        }
+    }
+
+    private suspend fun playSpringAnimation(
+        unconsumedDelta: Float,
+        initialVelocity: Float,
+        flingFromOverscroll: Boolean
+    ): Float {
+        val initialValue = overscrollOffset.toFloat() + unconsumedDelta
+
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Do we really need empty lines in such places? Not sure if it's under some styling rules

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+            playSpringAnimation(0f, velocity, true).toVelocity()
+        } else {
+            initialVelocity
+        }
+    }
+
+    private suspend fun playSpringAnimation(
+        unconsumedDelta: Float,
+        initialVelocity: Float,
+        flingFromOverscroll: Boolean
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: There is an opinion that such flag parameters are bad. It's more discussion topic, I'm not going to block anything because of it

> Robert C. Martin's Clean Code Tip 12: Eliminate Boolean Arguments
> Boolean arguments loudly declare that the function does more than one thing. They are confusing and should be eliminated.

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 11 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt:618]: Done

File: compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt (hunk 10 above)
Comment (elijah-semyonov) [reply compose/animation/animation-core/src/commonMain/kotlin/androidx/compose/animation/core/CupertionScrollDecayAnimationSpec.kt]: Done

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 11 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt:618]: Reverted, it's used from another class in the same file.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 12 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: `Density` contains `fontScale` and `density`. The calculations only depend on the second.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 13 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Done

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 14 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Done

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 15 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Done

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 16 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Done

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 18 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: That's the same way `ScrollingLogic` does it. It depends on internal class state and will make all calls, which are only used internally by this class more verbose.

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: You mean `Velocity` explicitly? Or all other conversions as well?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 19 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: A couple of those seem excessive, removed.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 20 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Agree, refactored

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 17 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: I double-checked how Offset is constructed before passed here: it is generated from 0f literal, so I just replaced those with strict 0f comparison.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt (hunk 18 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt]: Moved Velocity<->Offset to file scope.

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt (hunk 11 above)
Comment (MatkovIvan) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt:618]: internal or private extension (file scope) then

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/Scrollable.kt:618]: Class is already private and not visible beyond the file

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+ * Opt out of the Cupertino overscroll behavior (rubber banding and spring effect).
+ *
+ * This method sets the `isOptOut` flag to true, indicating that the Cupertino overscroll behavior should be disabled.
+ * By default, the Cupertino overscroll behavior is enabled on iOS.
+ *
+ * This method should be called before any @Composable function using this effect is executed
+ * (sometime during app start up).
+ */
+@OptIn(ExperimentalFoundationApi::class)
+fun optOutOfCupertinoOverscroll() {
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt]: We already have `ScrollConfig` class. Did you consider to add it there?

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+     */
+    private var direction: CupertinoOverscrollDirection = CupertinoOverscrollDirection.UNKNOWN
+
+    private val reverseHorizontal =
+        when (layoutDirection) {
+            LayoutDirection.Ltr -> false
+            LayoutDirection.Rtl -> true
+        }
+
+
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt]: Extra line

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt]: Removed

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt (hunk 21 above)
Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt]: Did a draft of it, have a look

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt [hunk 23]
```diff
@@ -27,9 +28,22 @@ import androidx.compose.ui.unit.IntSize
...
-internal actual fun platformScrollConfig(): ScrollConfig = UiKitConfig
+internal actual fun platformScrollConfig(): ScrollConfig = UiKitScrollConfig
+
+/**
+ * Opt out of the Cupertino overscroll behavior (rubber banding and spring effect).
+ *
+ * This method should be called before any @Composable function using this effect is executed
+ * (so as early as possible, e.g. during app start up).
+ */
+@OptIn(ExperimentalFoundationApi::class)
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt]: `OptIn` is about enabling using of experimental API, but not marking it as it

Comment (MatkovIvan) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt]:
```suggestion
@ExperimentalFoundationApi
```

Comment (elijah-semyonov) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt]: Fixed

**Review by MatkovIvan — APPROVED** [review]
